alpenhorn/update.py

```diff
@@ -36,86 +39,291 @@
 done_transport_this_cycle = False
 
 
-def update_loop(host):
-    """Loop over nodes performing any updates needed."""
+def update_loop(
+    host: str, queue: queue.FairMultiFIFOQueue, pool: WorkerPool | EmptyPool
+) -> None:
+    """Main loop of alepnhornd.
+
+    This is the main update loop for the alpenhorn daemon.
+
+    Parameters
+    ----------
+    host : string
+        the name of the host we're running on (must match the value in the DB)
+    queue : FairMultiFIFOQueue
+        the task manager
+    pool : WorkerPool
+        the pool of worker threads (may be empty)
+
+    The daemon cycles through the update loop until it is terminated in
+    one of three ways:
+
+    - receiving SIGINT (AKA KeyboardInterrupt).  This causes a clean exit.
+    - a global abort caused by an uncaught exception in a worker thread.  This
+        causes a clean exit.
+    - a crash due to an uncaught exception in the main thread.  This does _not_
+        cause a clean exit.
+
+    During a clean exit, alpenhornd will try to finish in-progress tasks before
+    shutting down.
+    """
     global done_transport_this_cycle
 
-    while True:
+    while not global_abort.is_set():
         loop_start = time.time()
         done_transport_this_cycle = False
 
-        # Iterate over nodes and perform each update (perform a new query
-        # each time in case we get a new node, e.g. transport disk)
-        for node in st.StorageNode.select().where(st.StorageNode.host == host):
-            update_node(node)
+        # Used to remember things about active groups and nodes on this host
+        groups = dict()
+        updated_nodes = list()
+
+        # Iterate over nodes first
+        #
```

Comment:
```suggestion
```

Reply:
I prefer having the blank line.  It acts as a paragraph break for the comment.  (Not that this is a terribly long comment, though.)